src/Sarif/Readers/AnnotatedCodeLocationIdConverter.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Globalization;
+using Newtonsoft.Json;
+
+namespace Microsoft.CodeAnalysis.Sarif.Readers
+{
+    public class AnnotatedCodeLocationIdConverter : JsonConverter
+    {
+        public override bool CanConvert(Type objectType)
+        {
+            return objectType == typeof(string) || objectType == typeof(long);
+        }
+
+        public override object ReadJson(JsonReader reader, Type objectType, object existingValue, JsonSerializer serializer)
+        {
+            int intValue;
+            object value = reader.Value;
+
+            // We have to look at the type of the object returned by reader.Value, rather
+            // than looking at objectType, because Newtonsoft does something a bit odd
+            // when the input is a string that parses as a valid integer, such as "1". In
+            // that case, objectType is Int32, but the object returned by reader.Value is
+            // a string!
+            Type valueType = value.GetType();
+
+            if (valueType == typeof(string))
+            {
+                var stringValue = (string)value;
+                if (!int.TryParse((string)value, out intValue))
+                {
+                    throw new ArgumentOutOfRangeException(
```

Comment:
The CLR exception doesn't say what I want. It just tells you that the thing you tried to parse isn't an integer. My error message (which I use again below, when I range-test the resulting integer) is:

    '{0}' is not a valid value for the id property of an annotatedCodeLocation object. The value must be a positive integer or the string representation of a positive integer.

This is fully informative, and applies whether the input was a malformed string ("x1") or an out-of-range integer (-1).

---
In reply to: [65252923](https://github.com/Microsoft/sarif-sdk/pull/316#discussion_r65252923) [](ancestors = 65252923)